crates/core/tedge_actors/src/message_boxes.rs

```diff
@@ -0,0 +1,265 @@
+use crate::Builder;
+use crate::ChannelError;
+use crate::DynSender;
+use crate::Message;
+use crate::MessageBoxConnector;
+use crate::SimpleMessageBoxBuilder;
+use async_trait::async_trait;
+use futures::channel::mpsc;
+use futures::StreamExt;
+use log::debug;
+use log::info;
+
+/// A message box used by an actor to collect all its input and forward its output
+///
+/// This message box can be seen as two streams of messages,
+/// - inputs sent to the actor and stored in the message box awaiting to be processed,
+/// - outputs sent by the actor and forwarded to the message boxes of other actors.
+///
+/// ```logical-view
+///                                      +------+
+/// input_sender: DynSender<Input> ----->| Box  |----> output_sender: DynSender<Input>
+///                                      +------+
+/// ```
+///
+/// Under the hood, a `MessageBox` implementation can use
+/// - several input channels to await messages from specific peers
+///   .e.g. awaiting a response from an HTTP actor
+///    and ignoring the kind of events till a response or a timeout has been received.
+/// - several output channels to send messages to specific peers.
+/// - provide helper function that combine internal channels.
+#[async_trait]
+pub trait MessageBox: 'static + Sized + Send + Sync {
+    /// Type of input messages the actor consumes
+    type Input: Message;
+
+    /// Type of output messages the actor produces
+    type Output: Message;
+
+    /// Return the next available input message if any
+    ///
+    /// Await for a message if there is not message yet.
+    /// Return `None` if no more message can be received because all the senders have been dropped.
+    async fn recv(&mut self) -> Option<Self::Input>;
+
+    /// Send an output message.
+    ///
+    /// Fail if there is no more receiver expecting these messages.
+    async fn send(&mut self, message: Self::Output) -> Result<(), ChannelError>;
+
+    /// Turn on/off logging of input and output messages
+    fn turn_logging_on(&mut self, on: bool);
+
+    /// Name of the associated actor
+    fn name(&self) -> &str;
+
+    /// Log an input message just after reception, before processing it.
+    fn log_input(&self, message: &Self::Input) {
```

Comment:
Instead of having these functions in this trait and expecting the actor author to explicitly use this, it would have been nice if this piece of code could be injected implicitly into the `send` and `recv` calls. Are there any Rust macro tricks for such code generation? Else we could have this function as the actual `send` function, provided as a default implementation, which is used by the actors and rename the current `send` and `recv` functions as `raw_send` and `raw_recv` which are trait functions to be implemented by the developer.

Reply:
I agree that this can be improved.

A simple solution is, as your propose, to provide `send` and `recv` default implementations that add logging to the `raw_send` and `raw_recv` methods provided by the message box. But  these methods `send` and `recv` are in practice not systematically used by the actors as noticed above.

That said a key responsibility of this `log_input` method is to standardize the log format. One must avoid each actor implementation coming with its own format.

Reply:
This will be addressed as a follow-up task.